README.md

```diff
@@ -1 +1,340 @@
+### 3.3 Varsig Body
+
+The varsig body MUST consist of zero or more segments, and MUST be defined by the signature algorithm.
+
+Some examples include:
+
+* Raw signature bytes only
+* CID of [DKIM] certification transparency record, and raw signature bytes
+* Hash algorithm multicodec prefix, signature counter, HMAC, and raw signature bytes
+
+# 4 Registry of Common Signature Algorithms
+
+Below are a few common signature headers and their fields.
+
+## 4.1 RSA
+
+RSASSA-PKCS #1 v1.5 signatures MUST include the following segments:
+
+``` abnf
+rsa-varsig-header = 0x1205 ; RSASSA-PKCS #1 v1.5
+rsa-hash-algorithm = unsigned-varint
+signature-byte-length = unsigned-varint
+sig-bytes = *OCTET
+```
+
+### 4.1.1 Example: RS256
+
+| Segment              | Hexadecimal | Unsigned Varint | Comment                                 | 
+|----------------------|-------------|-----------------|-----------------------------------------|
+| `rsa-varsig-header`  | `0x1205`    | `0x8524`        | RSASSA-PKCS #1 v1.5 [multicodec] prefix |
+| `rsa-hash-algorithm` | `0x12`      | `0x12`          | SHA2-256 [multicodec] prefix            |
+
+### 4.1.3 Example: RS512
+
+| Segment              | Hexadecimal | Unsigned Varint | Comment                                 | 
+|----------------------|-------------|-----------------|-----------------------------------------|
+| `rsa-varsig-header`  | `0x1205`    | `0x8524`        | RSASSA-PKCS #1 v1.5 [multicodec] prefix |
+| `rsa-hash-algorithm` | `0x13`      | `0x13`          | SHA2-512 [multicodec] prefix            |
+
+## 4.2 Ed25519
+
+``` abnf
+ed25519-varsig-header = 0xed ; Ed25519 multicodec prefix
+sig-bytes = 32(OCTET)
+```
+
+| Segment                 | Hexadecimal | Unsigned Varint | Comment                         | 
+|-------------------------|-------------|-----------------|---------------------------------|
+| `ed25519-varsig-header` | `0xed`      | `0xed01`        | Ed25519 key [multicodec] prefix |
+
+## 4.3 ECDSA
+
+ECDSA defines a general mechanism over many elliptic curves. Here are a few examples encoded as varsig:
+
+### 4.3.1 Example: ES256
+
+``` abnf
+es256-varsig-header = 0x1200 ; P-256 multicodec prefix
+sig-bytes = 64(OCTET)
+```
+
+| Segment                | Hexadecimal | Unsigned Varint | Comment                   | 
+|------------------------|-------------|-----------------|---------------------------|
+| `es256-varsig-header`  | `0x1200`    | `0x8024`        | P-256 [multicodec] prefix |
+
+### 4.3.1 Example: ES512
+
+``` abnf
+es512-varsig-header = 0x1202 ; P-521 multicodec prefix
+sig-bytes = 128(OCTET)
+```
+
+| Segment                | Hexadecimal | Unsigned Varint | Comment                   | 
+|------------------------|-------------|-----------------|---------------------------|
+| `es256-varsig-header`  | `0x1202`    | `0x8224`        | P-521 [multicodec] prefix |
+
+### 4.3.1 Example: ES256K
+
+``` abnf
+es256k-varsig-header = 0xe7 ; secp256k1 multicodec prefix
+sig-bytes = 64(OCTET)
```

Comment:
Don't we need to specify the hash alg here?
For `secp256k1` specifically, JOSE uses sha256, while ethereum wallets produce signatures with keccak256.

The same might be needed for other ECDSA algs as well?

Reply:
Other ECDSAs do make sense; we can add the hash parameter here. Do Ethereum wallets use e.g. EIP-191, or do they just sign the data directly?

Which opens the question about how we want to express things like EIP-191. Here's two options:


Per the 191 spec, I _think_ that we need to support all 3 varints under EIP-191... otherwise we can break them out into their own codes.

``` abnf
eip-191-varsig-header = 0xd191 ; EIP-191
eip-191-version = unsigned-varint
sig-bytes = 64(OCTET)
```

We could also nest the `Ethereum message` prefix from the signature type:

``` abnf
eip-191-signature = eip-191-header nested-varsig
eip-191-varsig-header = 0xd191 ; EIP-191
eip-191-version = unsigned-varint
nested-varsig = varsig
```

Thoughts?

Reply:
I went with the first one, but can change it if needed

Reply:
I also updated it to `0xe191`. In multicodec we had used `0xd191` we wanted to namespace all signatures into `0xd_`.

Reply:
This actually raises a different question for me. If I have an IPLD struct, what do I hash?

Suppose I encounter an object with the following schema,
```ipldsch
type Data struct {
  field1 Bytes
  field2 String
  sig Varsig
}

and the varsig is `varint(0x34) varint(0xed) sig-bytes`, e.g. an ED25519 signature, I know I should use `sha256`, but over what?

Reply:
Yeah it's a good question!

A previous iteration tried to include information about how the data was formatted before hashing (and thus including fields etc). However, to hopefully limit the scope of this PR and make it easier to get approval, Irakli suggested that we remove how the payload is formatted out of this spec (unless someone wants to make a case for including it again).

Your earlier suggestion of signing over the CID is a good option for a huge number of use cases (that's mentioned in `2.1`)! It doesn't work for things like JWTs and introduces pointer indirection, so the ability decide on what's being signed over seems like a good idea. This does mean that the exact signing methodology is left ambiguous from varsig itself, and formats that use varsig need to specify what's being signed over. We could always make a varsig (or multiformat) that does just that, so the spec as proposed here is not a one-way door by any means :)

Reply:
A bit of a tangent, but I also would love it if IPLD Schema added a few features like subtyping or splats, which would make it easier to describe that fields get signed over without introducing indirection. For example with sunbtyping:


```haskell
type Bar :< Foo struct {
    ...Foo  
    sig Varsig -- Signed Foo fields
}
```

But we don't have an established concept of subtyping in IPLD schema today

Reply:
I appreciate the effort to limit scope. However, I wonder what the use case of varsig is without encoding information? I thought the main purpose of varsig was to create a self-describing signature scheme?

For CACAOv3 having a way to know how to encode the data in order to verify the signature is actually quite important. Without it it's not necessarily trivial to determine if the payload is a UCAN or a SIWx (though it could be done by inspecting the object).

If we think limiting scope and adding extensions in the future is the right path, I'd like to better understand how a varsig without encoding info would be used!

Reply:
My argument had been that `sig-bytes` should use their own multiformat to describe how to derive bytes from the source data. If not specified it is `raw` bytes.

I also was making an argument that not every single signature would need to specify encoding which is why it's own multiformat makes sense.

I'm pretty sure that capturing every possible parameter in the tag is going to fail, so I suggest instead make VarSig one layer and let each variant define it's own thing and parameters as needed.

Reply:
@oed is this more in line with what you were thinking? I had this in a previous version, and would be happy to put it back if it solves problems for you!

```
                             ; vvvvvvvvvvvvvvvv
varsig = multibase-prefix %x34 content-encoding varsig-header varsig-body
multibase-prefix = ALPHA ; Multibase
content-encoding = unsigned-varint ; e.g. DAG-CBOR, raw bytes, etc
varsig-header = unsigned-varint ; Usually the public key code from Multicodec
varsig-body = *OCTET; Zero or more segments required by the kind of varsig (e.g. raw bytes, hash algorithm, etc)
```

Reply:
I had a call with @Gozala (about some other stuff) and we chatted about the above a bit. 

# Irakli's Preferred Version

```abnf
;; Irakli's Preference
varsig = multibase-prefix %x34 varsig-header varsig-body

;; Separate spec
content-info = 0xFF encoding-header encoding-body
```

Here, the signature bytes would be nested inside the content-encoding payload. For example:

```abnf
rsa-varsig = rsa-varsig-header rsa-hash-algorithm signature-byte-length content-info

;; For example, using brackets to show grouping
0x34 (0x1205 0x12 0x0100 (0xFF 0x71 signature-bytes))
     ^                   ^
     |                   |
     |          nested-content-encoding    
     |
  rsa-sig
```

The reason for this style is that you can endlessly nest these, and each segment acts as its own multiformat with its own number of segments. The downside is that if you want to update a segmant, you need to change some deeply nested structure.

# Wrapping

Here's basically the opposite version which works by adding info as needed at the top level (which could lead to a lot of varsig types but keeps the "core" signature info together with the bytes)

```anbf
0x34 (0xFF 0x71 (0x1205 0x12 0x0100 signature-bytes)
     ^          ^
     |          |
     |       rsa-sig
     |
 wrapped-content-encoding
```

# Tree Structure

Arguably, we know that at minimum we want the encoding, so including at the top level rather than pushing it into every variation:

```abnf
varsig = multibase-prefix %x34 encoding-info varsig-header varsig-body sig-bytes
```

```anbf
0x34 (0x71) (0x1205 0x12 0x0100) (signature-bytes)
     ^      ^                    ^
     |      |                    |
     |   rsa-sig            raw-rsa-signature
     |
 wrapped-content-encoding
```

Reply:
> My argument had been that sig-bytes should use their own multiformat to describe how to derive bytes from the source data. If not specified it is raw bytes.

How do I go from the IPLD object to `raw` bytes if it's not specified? I think storing the encoding inside of the `sig-bytes` seems fine, but the examples in the spec needs to demonstrate this if so.

@expede Thanks for the examples. I'm kind of fine either way as long as we specify the encoding. 

I would also like to know if you see any specific use case that doesn't require encoding @Gozala ?

Finally a question around the content encoding itself. Assume we have the following object IPLD object:
```
type Foo struct {
  bar String
  sig Varsig
}
```

My assumption would be that we remove the `sig` field (or whatever field contained the varisg, then do the encoding? Is this other peoples assumption as well, or is there a different approach?

Reply:
> Is this other peoples assumption as well, or is there a different approach?

This is also my assumpition. It's not fully clear how you would sign over just the `bar` string (but not the key) — I would assume that this is by using the `0x55 raw data` encoding, or introduce a "forget about the keys just teh values stringjoined" encoding prefix. Alternately pushing this decision into the schema layer frees up this choice (like the extended IPLD subtyping example earlier in this thread).

Reply:
> @expede Thanks for the examples. I'm kind of fine either way as long as we specify the encoding.

Okay let's go with Irakli's preference then, since neither of us have super string feelings against 👍

Reply:
I went ahead and updated the spec with `encoding-info`. I embedded the table right in this spec; we can always break this out at a later date. @oed it's been a _while_ since I worked with Ethereum signatures, can you confirm that I haven't missed any nuances?